test/config-luatest/config_info_hierarchy_test.lua

```diff
@@ -0,0 +1,153 @@
+g.test_2_11_1 = function(g)
+    local config = cbuilder:new()
+        :use_group('g-001')
+        :use_replicaset('r-001')
+        :add_instance('i-001', {
+            -- We have to set the UUIDs, because a safety check
+            -- requires to verify the configuration <-> database
+            -- correspondence either by names or by UUIDs, and
+            -- there are no names in the given snapshot.
+            database = {
+                instance_uuid = '0679afea-8275-4028-bf7a-a83adc18b844',
+                replicaset_uuid = '02140ab2-0d0e-421e-832e-d79024392a24',
+            },
+        })
+        :config()
+
+    -- Copy the 2.11.1 snapshot.
+    --
+    -- Can't use the `datadir` luatest.server option, because its
+    -- content is copied to the `workdir` directory, while the
+    -- cluster helper starts servers in a temporary directory
+    -- using the `chdir` option instead of `workdir`.
+    local snap = 'test/box-luatest/upgrade/2.11.1/00000000000000000000.snap'
+    local cluster = cluster.new(g, config)
+    local data_dir = fio.pathjoin(cluster._dir, 'var/lib/i-001')
+    fio.mktree(data_dir)
+    assert(fio.copyfile(snap, data_dir))
+
+    -- Add permissions (won't needed after gh-10539).
+    --
+    -- Otherwise it is not possible to connect to the instance
+    -- from the test and the ready check fails on the startup.
+    --
+    -- NB: If this sequence of steps is failed in the middle (so
+    -- we don't reach g.it:close()), the g.it handle is collected
+    -- by GC and the process is killed. At worst, it is performed
+    -- on luatest termination.
+    g.it = it.new()
+    g.it:roundtrip("fio = require('fio')")
+    g.it:roundtrip(('fio.chdir(%q)'):format(data_dir))
+    g.it:roundtrip('box.cfg()')
+    g.it:roundtrip("box.schema.user.create('client', {password = 'secret'})")
+    g.it:roundtrip("box.schema.user.grant('client', 'super')")
+    g.it:close()
+    g.it = nil
+
+    -- Start with 2.11.1 snapshot.
+    cluster:start()
+
+    -- Verify a test prerequisite: the names are not written to
+    -- the database.
+    cluster['i-001']:exec(function()
+        t.assert_equals(box.info.name, box.NULL)
+        t.assert_equals(box.info.replicaset.name, box.NULL)
+    end)
+
+    -- Verify that the configured names are shown.
+    --
+    -- All the previous steps are setup and we do it just to call
+    -- this little function!
+    assert_configured_names(cluster['i-001'])
+end
+
+-- Verify that there are no instance/replicaset/group names in
+-- config:info() if tarantool is started from a Lua script,
+-- without the declarative configuration.
+g.test_script = function(g)
+    g.server:exec(function()
+        local config = require('config')
+
+        t.assert_equals(config:info().hierarchy, {})
+        t.assert_equals(config:info('v1').hierarchy, {})
+        t.assert_equals(config:info('v2').hierarchy, {})
+        t.assert_equals(box.info.config.hierarchy, {})
```

Comment:
Nit: such `assert_config_info_hierarchy(g.server, {})` could be used here, for example.

Reply:
Good idea. I've implemented it.